Do not run version check if generating shell completion and stdout isn't terminal

corpulentcoffee opened the pull request:
In addition to suppressing the version check if running in CI or with the machine flag, suppress it if stdout is being piped or redirected while using one of the shell completion commands.

Fixes #79202.

Happy to amend this to use one of the other ways proposed in the original issue.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], ~~including [Features we expect every widget to implement].~~
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] ~~I updated/added relevant documentation (doc comments with `///`).~~
- [x] I added new tests to check the change I am making or feature I am adding, ~~or Hixie said the PR is test-exempt.~~
- [x] All existing and new tests are passing

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

jonahwilliams commented:
We're not quite sure we should turn this off globally, perhaps a more targeted change to the shell completion commands first?

jonahwilliams commented:
For example: what if the tool were piping output to a log file, we'd probably still want to see it.

corpulentcoffee committed:
Allow naming DummyFlutterCommand in runner tests

This enables the testing of command-dependent code paths in the command
runner.

corpulentcoffee committed:
Add failing test for redirected shell completion

Specifically, we should not do an update check if the user is generating
shell completion and redirecting the terminal output somewhere else.

corpulentcoffee committed:
Skip update check if redirecting shell completion

Collectively refer to CI, the machine flag, and redirecting shell
completion output as being a "machine" thing, and don't do the update
check under those circumstances.

corpulentcoffee commented:
> perhaps a more targeted change to the shell completion commands first

Works for me -- because the check now depends on the name of the command, I adjusted `DummyFlutterCommand` to allow test cases to specify a name other than `'dummy'`.

jonahwilliams reviewed (approved):
LGTM
